crates/vault-registry/src/slash.rs

```diff
@@ -0,0 +1,395 @@
+pub enum ArithmeticError {
+    TryIntoIntError,
+    ArithmeticUnderflow,
+    ArithmeticOverflow,
+}
+
+fn inner_to_backing<Backing: TryFrom<u128>, Inner: FixedPointOperand>(x: Inner) -> Result<Backing, ArithmeticError> {
+    let y = UniqueSaturatedInto::<u128>::unique_saturated_into(x);
+    TryInto::<Backing>::try_into(y).map_err(|_| ArithmeticError::TryIntoIntError)
+}
+
+fn backing_to_inner<Backing: TryInto<u128>, Inner: FixedPointOperand>(x: Backing) -> Result<Inner, ArithmeticError> {
+    let y = TryInto::<u128>::try_into(x).map_err(|_| ArithmeticError::TryIntoIntError)?;
+    TryInto::<Inner>::try_into(y).map_err(|_| ArithmeticError::TryIntoIntError)
+}
+
+fn backing_to_fixed<Backing: TryInto<u128>, SignedFixedPoint: FixedPointNumber>(
+    x: Backing,
+) -> Result<SignedFixedPoint, ArithmeticError> {
+    let signed_fixed_point =
+        SignedFixedPoint::checked_from_integer(backing_to_inner::<Backing, SignedFixedPoint::Inner>(x)?)
+            .ok_or(ArithmeticError::TryIntoIntError)?;
+    Ok(signed_fixed_point)
+}
+
+pub trait Collateral<Backing, SignedFixedPoint, E: From<ArithmeticError>> {
+    fn get_slash_per_token(&self) -> SignedFixedPoint;
+
+    fn get_collateral(&self) -> Backing;
+
+    fn mut_collateral<F>(&mut self, func: F) -> Result<(), E>
+    where
+        F: Fn(&mut Backing) -> Result<(), E>;
+
+    fn get_total_collateral(&self) -> Backing;
+
+    fn mut_total_collateral<F>(&mut self, func: F) -> Result<(), E>
+    where
+        F: Fn(&mut Backing) -> Result<(), E>;
+
+    fn get_slash_tally(&self) -> SignedFixedPoint;
+
+    fn mut_slash_tally<F>(&mut self, func: F) -> Result<(), E>
+    where
+        F: Fn(&mut SignedFixedPoint) -> Result<(), E>;
+}
+
+pub trait Slashable<
+    Backing: TryInto<u128> + TryFrom<u128>,
+    SignedFixedPoint: FixedPointNumber,
+    E: From<ArithmeticError>,
+>: Collateral<Backing, SignedFixedPoint, E>
+{
+    fn mut_slash_per_token<F>(&mut self, func: F) -> Result<(), E>
+    where
+        F: Fn(&mut SignedFixedPoint) -> Result<(), E>;
+
+    fn compute_total_collateral(&self) -> Result<Backing, E> {
+        let total_collateral = self.get_total_collateral();
+        let total_collateral_as_fixed = backing_to_fixed::<Backing, SignedFixedPoint>(total_collateral)?;
+        let slash_per_token = self.get_slash_per_token();
+        let to_slash = total_collateral_as_fixed
+            .checked_mul(&slash_per_token)
+            .ok_or(ArithmeticError::ArithmeticOverflow)?;
+        inner_to_backing::<Backing, SignedFixedPoint::Inner>(
+            total_collateral_as_fixed
+                .checked_sub(&to_slash)
+                .ok_or(ArithmeticError::ArithmeticUnderflow)?
+                .into_inner()
+                .checked_div(&SignedFixedPoint::accuracy())
+                .ok_or(ArithmeticError::ArithmeticUnderflow)?,
+        )
+        .map_err(Into::into)
+    }
+
+    fn slash_collateral(&mut self, amount: Backing) -> Result<(), E> {
```

Comment:
It uses the same naming as the module functionality also for `try_deposit` and `try_withdraw`. I think it could be confusing if they are all named differently. Do you have another suggestion?

Reply:
The other slash_collateral is a bit of a misnomer, since it can also transfer to/from free_balance. Maybe you can rename that one to `transfer_funds` or something

Reply:
Yeah that makes sense, though I would do that in a separate PR since it touches more files.